### PR TITLE
add transaction overview tikv document

### LINKED ISSUE #150: Write down the first version of Transaction on TiKV

### PR DESCRIPTION
**Pull request by cfzjywxk**
### What issue does this PR solve?

- close #150

### What is changed:

Add the overview document for transaction engine, the tikv part.

### CODE REVIEW

**Comment by cfzjywxk** [comment]
@you06 
Seems the related issue is not created.

**Review by tisonkun — COMMENTED** [review]
Thanks for you contribution! You may try to update `SUMMARY.md` also so that we can list this file on the table of content. You can run `mdbook serve` locally to see whether the render version is as you wish.

File: src/understand-tidb/transaction_kv.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    // The max commit ts is reserved for limiting the commit ts of 1PC or async commit, which can be used to avoid
+    // inconsistency with schema change.
+    uint64 max_commit_ts = 14;
+}
+```
+
+The `mutations` are changes made by the transaction, `start_version` is the transaction identifier fetched from PD, `for_update_ts` is used by the pessimistic transactions. The `try_one_pc` field is used if the transaction is committed using `one-phase` protocol, the `use_async_commit` and `secondaries` will be used if the transaction is committed in the `async-commit` mode.
+
+Besides `prewrite` request, there are some other important request types:
+- `pessimistic_lock` [request](https://github.com/pingcap/kvproto/blob/0f5764a128ad77ccf0a5b0ce0d6e2bfa50a108ce/proto/kvrpcpb.proto#L125) is used to lock a key. Note pessimistic locking happens in the transaction execution phase, for example a `select for update` statment will need to pessimistically lock the correspond rows.
```
Comment (sticnarf) [inline src/understand-tidb/transaction_kv.md:49]:
```suggestion
- `pessimistic_lock` [request](https://github.com/pingcap/kvproto/blob/0f5764a128ad77ccf0a5b0ce0d6e2bfa50a108ce/proto/kvrpcpb.proto#L125) is used to lock a key. Note pessimistic locking happens in the transaction execution phase, for example a `select for update` statement will need to pessimistically lock the corresponding rows.
```

File: src/understand-tidb/transaction_kv.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+If the transaction command processing in the transaction scheduler is successful, the generated transaction writes will be written into the raft log engine by the region leaders in `raftStore`(raftStore will be introduced in other documents in details). The work flow is like this:
+
+![raftStore](../img/raftStore.png)
+
+The writes generated by transaction commands will be sent to the raft peer message task queue first, then the raft batch system will poll each raft peer and handle these requests in the raft thread-pool. After all the raft logs are persisted on majority raft group members, they are regarded as `commit`. Then the correspond apply task be delivered to the apply worker pool to apply the actual write contents to the storage engine, after that the transaction command processing is considered successful and the callback will be invoked to response `OK` results to the RPC client.
+
+
+# Transaction Record In TiKV
+
+In TiDB a transaction is considered committed only if its primary key lock is committed succesfully(If async commit protocol is not used). The acutal key and value written into storage engine is in the following format:
```
Comment (sticnarf) [inline src/understand-tidb/transaction_kv.md:76]:
```suggestion
In TiDB a transaction is considered committed only if its primary key lock is committed succesfully (if async commit protocol is not used). The actual key and value written into storage engine is in the following format:
```

File: src/understand-tidb/transaction_kv.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+Actually there is no special mechanism to recover the failed transctions, the recovery is done by other concurrent conflict transactions, or the conflict transactions will help decide the actual states of the failed tranactions. The lock resolve process will be triggered if current ongoing transaction encounters locks of other transactions doing reads or writes. The whole resolve process will be introduced in other documents in details.
+
+# Transaction Optimizations
+
+Normally the transaction commit will need two pahses, the `prewrite` phase and `commit` phase. Under certain circumstances transaction commit could be done in a single phase for example the generated transaction mutations could be processed by a single region leader. This optimization is called one-phase commit in TiDB. The final transaction status is determined by the commit status of the primary key lock, so the reponse to the client has to wait until the primary key commit has finished. This wait could be saved using the `async-commit` protocol so the latency of commit processing could be reduced. They will both be introduced other documents in details.
```
Comment (sticnarf) [inline src/understand-tidb/transaction_kv.md:95]:
```suggestion
Normally the transaction commit will need two pahses, the `prewrite` phase and `commit` phase. Under certain circumstances transaction commit could be done in a single phase for example the generated transaction mutations could be processed by a single region leader. This optimization is called one-phase commit in TiDB.

The final transaction status is determined by the commit status of the primary key lock, so the reponse to the client has to wait until the primary key commit has finished. This wait could be saved using the `async-commit` protocol so the latency of commit processing could be reduced.

They will both be introduced other documents in details.
```

Put each optimization in separate paragraphs to make it clearer.

File: src/understand-tidb/transaction_kv.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+# Summary
+
+This section talks about the brief steps of tranaction processing in the TiKV part, and related interfaces, implementations and optimizations.
```
Comment (sticnarf) [inline src/understand-tidb/transaction_kv.md:99]:
```suggestion
This section talks about the brief steps of transaction processing in the TiKV part, and related interfaces, implementations and optimizations.
```

**Comment by you06** [comment]
> @you06
> Seems the related issue is not created.

This chapter may transfer to [tikv-dev-guide](https://github.com/tisonkun/tikv-dev-guide)? @tisonkun what's your idea?

**Comment by tisonkun** [comment]
> > @you06
> > Seems the related issue is not created.
> 
> This chapter may transfer to [tikv-dev-guide](https://github.com/tisonkun/tikv-dev-guide)? @tisonkun what's your idea?

It sounds good. We can setup cross references for each other on topic cross boundary.

**Comment by sticnarf** [comment]
Sounds like my async commit doc also needs to be split to the two repos.

**Comment by tisonkun** [comment]
> Sounds like my async commit doc also needs to be split to the two repos.

I don't mind to keep some overlap so that reader can get necessary information in one guide.

File: src/understand-tidb/transaction_kv.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+The future processing is done in the transaction scheduler thread-pool, usually there will be some tasks like conflict and constraint checks, write mutation generations. For example, the prewrite request processing will need to check if there is already a conflict [lock](https://github.com/tikv/tikv/blob/6be3893f7f787b04bf34d99d1369092404ab5cfc/src/storage/txn/actions/prewrite.rs#L45) or a conflict committed write [record](https://github.com/tikv/tikv/blob/6be3893f7f787b04bf34d99d1369092404ab5cfc/src/storage/txn/actions/prewrite.rs#L59).
```
Comment (you06) [inline src/understand-tidb/transaction_kv.md:61]:
```suggestion
The task will be processed as a [future](https://doc.rust-lang.org/stable/std/future/trait.Future.html). The future processing is done in the transaction scheduler thread-pool. Usually, there will be some tasks like conflict and constraint checks, write mutation generations. For example, the prewrite request processing will need to check if there is already a conflict [lock](https://github.com/tikv/tikv/blob/6be3893f7f787b04bf34d99d1369092404ab5cfc/src/storage/txn/actions/prewrite.rs#L45) or a conflict committed write [record](https://github.com/tikv/tikv/blob/6be3893f7f787b04bf34d99d1369092404ab5cfc/src/storage/txn/actions/prewrite.rs#L59).
```

Comment (you06) [reply src/understand-tidb/transaction_kv.md:61]: It takes me seconds thinking about what does "future" indicates here.

File: src/understand-tidb/transaction_kv.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+# Transaction Recovery
+
+In TiDB the transaction coordinator(in `tidb-server`) is stateless and it will not persist any information. If the transaction coordinator fails for example the `tidb-server` crashes, the transaction context in memory will get lost, and as the coordinator is gone the normal commit processing will stop. How to recover the transaction state and make a decision if it should commit or abort? 
+
+Actually there is no special mechanism to recover the failed transctions, the recovery is done by other concurrent conflict transactions, or the conflict transactions will help decide the actual states of the failed tranactions. The lock resolve process will be triggered if current ongoing transaction encounters locks of other transactions doing reads or writes. The whole resolve process will be introduced in other documents in details.
```
Comment (you06) [inline src/understand-tidb/transaction_kv.md:91]: The expression "failed transctions" is a little misunderstanding. Will "undetermined-status transactions" be more proper?

File: src/understand-tidb/transaction_kv.md [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+# Transaction Optimizations
+
+Normally the transaction commit will need two pahses, the `prewrite` phase and `commit` phase. Under certain circumstances transaction commit could be done in a single phase for example the generated transaction mutations could be processed by a single region leader. This optimization is called one-phase commit in TiDB.
```
Comment (you06) [inline src/understand-tidb/transaction_kv.md:95]:
```suggestion
Normally the transaction commit will need two phases, the `prewrite` phase and `commit` phase. Under certain circumstances transaction commit could be done in a single phase for example the generated transaction mutations could be processed by a single region leader. This optimization is called one-phase commit in TiDB.
```

File: src/understand-tidb/transaction_kv.md [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+The final transaction status is determined by the commit status of the primary key lock, so the reponse to the client has to wait until the primary key commit has finished. This wait could be saved using the `async-commit` protocol so the latency of commit processing could be reduced.
```
Comment (you06) [inline src/understand-tidb/transaction_kv.md:97]:
```suggestion
The final transaction status is determined by the commit status of the primary key lock, so the response to the client has to wait until the primary key commit has finished. This wait could be saved using the `async-commit` protocol so the latency of commit processing could be reduced.
```

**Review by tisonkun — APPROVED** [review]
LGTM. @you06 I think we can duplicate a bit this section in both Dev Guide and create cross references.

**Comment by cfzjywxk** [comment]
@tisonkun 
LGTM, thanks for the help~

**Comment by tisonkun** [comment]
@cfzjywxk thanks for your feedback. I'm waiting for @you06 's final check and then I can merge this PR.

**Comment by tisonkun** [comment]
Thanks for your contribution @cfzjywxk ! Merging...

**Pull request merged**